qubit-sdk/src/main/java/com/qubit/android/sdk/internal/placement/interactor/PlacementQueryAttributesBuilder.kt

```diff
@@ -0,0 +1,97 @@
+package com.qubit.android.sdk.internal.placement.interactor
+
+import com.google.gson.JsonArray
+import com.google.gson.JsonElement
+import com.google.gson.JsonObject
+import com.google.gson.JsonPrimitive
+import com.qubit.android.sdk.internal.placement.repository.PlacementAttributesRepository
+
+internal class PlacementQueryAttributesBuilder(
+    private val placementAttributesRepository: PlacementAttributesRepository
+) {
+
+  companion object {
+    internal const val VISITOR_ATTRIBUTE_KEY = "visitor"
+    internal const val USER_ATTRIBUTE_KEY = "user"
+    internal const val VIEW_ATTRIBUTE_KEY = "view"
+  }
+
+  internal fun buildJson(
+      deviceId: String,
+      customAttributes: JsonObject?
+  ): JsonObject {
+    val attributesJson = JsonObject()
+
+    // 1 visitor attributes should be set by SDK
+    attributesJson.add(VISITOR_ATTRIBUTE_KEY, buildVisitorAttributesJson(deviceId))
+
+    // 2 custom attributes passed by SDK user
+    if (customAttributes != null) {
+      customAttributes.remove(VISITOR_ATTRIBUTE_KEY)
+      for (key in customAttributes.keySet()) {
+        // user/view attributes should contain all fields from schema and nothing more
+        val customAttribute = prepareCustomAttribute(key, customAttributes.get(key))
+        attributesJson.add(key, customAttribute)
+      }
+    }
+
+    // 3 cached attributes - skipped if set by SDK user
+    val cachedAttributes = placementAttributesRepository.load()
+    for (key in cachedAttributes.keys) {
+      if (!attributesJson.has(key)) {
+        attributesJson.add(key, cachedAttributes[key])
+      }
+    }
+
+    // 4 if user/view attributes are missing, then we should add empty ones
+    if (!attributesJson.has(USER_ATTRIBUTE_KEY)) {
+      attributesJson.add(USER_ATTRIBUTE_KEY, buildEmptyUserAttribute())
+    }
+    if (!attributesJson.has(VIEW_ATTRIBUTE_KEY)) {
+      attributesJson.add(VIEW_ATTRIBUTE_KEY, buildEmptyViewAttribute())
+    }
+    return attributesJson
+  }
+
+  private fun prepareCustomAttribute(key: String, customValue: JsonElement): JsonElement {
```

Comment:
Invokation of this method is commented:
// user/view attributes should contain all fields from schema and nothing more

This is not quite fulfilled.
When key is "user", but customValue is not json object, it returns customValue which is not compatible with expected schema.

Reply:
Right. Fixed and added unit tests to verify that.